### PR TITLE
scheduler: reimplementation of timer scheduling

### PR DESCRIPTION
**Pull request by tlauda**
Refactors implementation of timer scheduling in order to get it work correctly:
- Adds scheduling mode to pipeline.
- Moves timer work from dw-dma to pipeline.
- Adds new operation avail_data_size to dw-dma and hd-dma.
- Reimplements dai and host copy methods.
- Reimplements dw-dma and hd-dma copy methods.

### CODE REVIEW

File: src/drivers/dw/dma.c [hunk 1]
```diff
@@ -1411,6 +1437,56 @@ static int dw_dma_remove(struct dma *dma)
 	return 0;
 }
 
+static int dw_dma_in_data_size(struct dma_chan_data *chan)
+{
+	int size = 0;
+	uint32_t read_ptr = chan->ptr_data.current_ptr;
+	uint32_t write_ptr = dw_read(chan->id.dma, DW_DAR(chan->id.channel));
+
+	size = (int)write_ptr - (int)read_ptr;
```
Comment (plbossart) [inline src/drivers/dw/dma.c]: isn't this what intptr_t was invented for?

File: src/include/sof/dma.h [hunk 2]
```diff
@@ -85,10 +85,13 @@
...
 #define DMA_COPY_PRELOAD	BIT(0)
 
+/* DMA preload threshold */
+#define DMA_PRELOAD_THRESHOLD(x)	((x) * 2)
+
 /* We will use this macro in cb handler to inform dma that
  * we need to stop the reload for special purpose
  */
-#define DMA_RELOAD_END	0
+#define DMA_RELOAD_END	0xFFFFFFFE
```
Comment (plbossart) [inline src/include/sof/dma.h]: this is odd.
The only related change I see is this:
		if (next.size == DMA_RELOAD_END) {
		if (next.size == DMA_RELOAD_END)

But it's not clear what impact the change for this define has? Is this intentional and does this belong in this timer-scheduling change or to a different patch?

File: src/include/sof/dma.h [hunk 3]
```diff
@@ -121,7 +121,7 @@ struct dma_sg_config {
 	uint32_t src_dev;
 	uint32_t dest_dev;
 	uint32_t cyclic;			/* circular buffer */
-	uint32_t timer_delay;	/* non zero if timer scheduled */
+	uint32_t timer;				/* timer scheduling */
```
Comment (lyakh) [inline src/include/sof/dma.h]: This one is really "bool" now, can we use that type here?

File: src/drivers/dw/dma.c [hunk 4]
```diff
@@ -282,7 +282,7 @@ struct dma_chan_data {
 	uint32_t cfg_lo;
 	uint32_t cfg_hi;
 	struct dma_id id;
-	uint32_t timer_delay;
+	uint32_t timer;
```
Comment (lyakh) [inline src/drivers/dw/dma.c]: Why are you renaming this? This is still referring to a time interval, right? Maybe it would be good to keep "timer_delay" to help differentiate it from the "bool timer" from struct dma_sg_config?

File: src/drivers/dw/dma.c [hunk 5]
```diff
@@ -702,7 +702,7 @@ static int dw_dma_set_config(struct dma *dma, int channel,
 
 	/* default channel config */
 	chan->direction = config->direction;
-	chan->timer_delay = config->timer_delay;
+	chan->timer = config->timer;
```
Comment (lyakh) [inline src/drivers/dw/dma.c]: Now this confuses me, aren't you assigning a boolean value to a time-interval variable? You really mean a delay of 1?

File: src/include/sof/audio/pipeline.h [hunk 6]
```diff
@@ -97,6 +101,7 @@ struct pipeline {
 	uint32_t status;		/* pipeline status */
 
 	/* scheduling */
+	uint32_t scheduling_mode;	/* pipeline scheduling mode */
```
Comment (lyakh) [inline src/include/sof/audio/pipeline.h]: enum?

File: src/drivers/dw/dma.c [hunk 7]
```diff
@@ -1234,10 +1190,12 @@ static void dw_dma_irq_handler(void *data)
 		return;
 	}
 
+#if DW_USE_HW_LLI
```
Comment (lyakh) [inline src/drivers/dw/dma.c]: strictly speaking adding these "#if's" is unrelated... Would be good to have a separate patch with a proper description for that

File: src/drivers/dw/dma.c [hunk 8]
```diff
@@ -945,6 +964,13 @@ static int dw_dma_set_config(struct dma *dma, int channel,
 	chan->status = COMP_STATE_PREPARE;
 	chan->lli_current = chan->lli;
 
+	/* initialize pointers */
+	chan->ptr_data.start_ptr = (chan->direction == DMA_DIR_MEM_TO_DEV) ?
```
Comment (lyakh) [inline src/drivers/dw/dma.c]: superfluous parentheses

File: src/drivers/dw/dma.c [hunk 9]
```diff
@@ -1411,6 +1437,56 @@ static int dw_dma_remove(struct dma *dma)
 	return 0;
 }
 
+static int dw_dma_in_data_size(struct dma_chan_data *chan)
+{
+	int size = 0;
```
Comment (lyakh) [inline src/drivers/dw/dma.c]: redundant initialisation. You could also move it two lines down and initialise with the correct value directly

File: src/drivers/dw/dma.c (hunk 1 above)
Comment (lyakh) [reply src/drivers/dw/dma.c]: @plbossart you mean for read_ptr and write_ptr? For size it would be rather ssize_t? Are type-casts really needed?

File: src/drivers/dw/dma.c [hunk 10]
```diff
@@ -1411,6 +1437,56 @@ static int dw_dma_remove(struct dma *dma)
...
+	size = (int)write_ptr - (int)read_ptr;
+	if (size < 0)
+		size += chan->ptr_data.buffer_bytes;
+
+	return size;
+}
+
+static int dw_dma_out_data_size(struct dma_chan_data *chan)
+{
+	int size = 0;
```
Comment (lyakh) [inline src/drivers/dw/dma.c]: ditto

File: src/drivers/dw/dma.c [hunk 11]
```diff
@@ -1411,6 +1437,56 @@ static int dw_dma_remove(struct dma *dma)
...
+
+	spin_lock_irq(&dma->lock, flags);
+
+	tracev_dwdma("dw-dma: %d channel %d avail_data_size",
+		     dma->plat_data.id, channel);
+
+	if (chan->status != COMP_STATE_ACTIVE)
+		goto out;
+
+	data_size = (chan->direction == DMA_DIR_MEM_TO_DEV) ?
```
Comment (lyakh) [inline src/drivers/dw/dma.c]: parentheses

File: src/drivers/intel/cavs/hda-dma.c [hunk 12]
```diff
@@ -1005,6 +1005,81 @@ static int hda_dma_remove(struct dma *dma)
 	return 0;
 }
 
+static int hda_dma_in_data_size(struct hda_chan_data *chan)
+{
+	int size = 0;
+	uint32_t read_ptr;
+	uint32_t write_ptr;
```
Comment (lyakh) [inline src/drivers/intel/cavs/hda-dma.c]: you could declare read_ptr and write_ptr inside the "if" clause. Alternatively you could invert the if like
```
if (!...)
	return 0;
...
return size;
```
which would unindent the whole function.

File: src/drivers/intel/cavs/hda-dma.c [hunk 13]
```diff
@@ -1005,6 +1005,81 @@ static int hda_dma_remove(struct dma *dma)
...
+
+	return size;
+}
+
+static int hda_dma_out_data_size(struct hda_chan_data *chan)
+{
+	int size = 0;
+	uint32_t status;
+	uint32_t read_ptr;
+	uint32_t write_ptr;
```
Comment (lyakh) [inline src/drivers/intel/cavs/hda-dma.c]: ditto

File: src/drivers/intel/cavs/hda-dma.c [hunk 14]
```diff
@@ -1005,6 +1005,81 @@ static int hda_dma_remove(struct dma *dma)
...
+{
+	struct dma_pdata *p = dma_get_drvdata(dma);
+	struct hda_chan_data *chan = &p->chan[channel];
+	int data_size = 0;
+	uint32_t flags;
+
+	if (channel >= HDA_DMA_MAX_CHANS) {
+		trace_hddma_error("hda-dmac: %d invalid channel %d",
+				  dma->plat_data.id, channel);
+		return data_size;
```
Comment (lyakh) [inline src/drivers/intel/cavs/hda-dma.c]: return 0;

File: src/drivers/intel/cavs/hda-dma.c [hunk 15]
```diff
@@ -1005,6 +1005,81 @@ static int hda_dma_remove(struct dma *dma)
...
+	}
+
+	spin_lock_irq(&dma->lock, flags);
+
+	tracev_hddma("hda-dmac: %d channel %d -> avail_data_size",
+		     dma->plat_data.id, channel);
```
Comment (lyakh) [inline src/drivers/intel/cavs/hda-dma.c]: does the trace really have to be under the spin-lock?

File: src/drivers/dw/dma.c (hunk 1 above)
Comment (dbaluta) [reply src/drivers/dw/dma.c]: @lyakh similar with int32_t there is intptr_t. The purpose of this pointer typedef is to be able to substract pointers.
It is similar with the cast in your patch but in a more fancy way.

https://github.com/thesofproject/newlib-xtensa/blob/xtensa/newlib/libc/include/stdint.h#L247

File: src/drivers/dw/dma.c [hunk 16]
```diff
@@ -1170,6 +1170,45 @@ static void dw_dma_process_block(struct dma_chan_data *chan,
...
+	if (chan->status != COMP_STATE_ACTIVE) {
+		trace_dwdma_error("dw-dma: %d channel %d not running",
+				  dma->plat_data.id, channel);
+
+		/* skip if channel is not running */
+		return 0;
+	}
+
+	if (chan->cb)
+		chan->cb(chan->cb_data, DMA_IRQ_TYPE_BLOCK, &next);
```
Comment (lyakh) [inline src/drivers/dw/dma.c]: Could you clarify this a bit - is this some special case? Isn't the callback usually called upon DMA completion?

File: src/drivers/dw/dma.c (hunk 1 above)
Comment (lyakh) [reply src/drivers/dw/dma.c]: Thanks, I know what intptr_t is for. I meant - shouldn't intptr_t be used for read_ptr and write_ptr, not for "size"

File: src/include/sof/dma.h (hunk 2 above)
Comment (tlauda) [reply src/include/sof/dma.h]: Getting current available data from DMA can be 0, which is valid in some situations e.g. work_queue tick happened right after starting DMA. Keeping DMA_RELOAD_END as 0 would stop DMA.

File: src/drivers/intel/cavs/hda-dma.c [hunk 17]
```diff
@@ -155,6 +155,7 @@ struct hda_chan_data {
 	uint32_t desc_count;
 	uint32_t desc_avail;
 	uint32_t direction;
+	uint32_t timer;
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/drivers/intel/cavs/hda-dma.c]: Changed to bool.

File: src/audio/pipeline.c [hunk 18]
```diff
@@ -77,14 +78,20 @@ struct pipeline *pipeline_new(struct sof_ipc_pipe_new *pipe_desc,
...
 	memcpy(&p->ipc_pipe, pipe_desc, sizeof(*pipe_desc));
 	p->scheduling_mode = pipe_desc->timer_delay ?
 		PPL_SCHED_TIMER_IRQ : PPL_SCHED_DMA_IRQ;
 
+	if (p->scheduling_mode == PPL_SCHED_DMA_IRQ) {
+		schedule_task_init(&p->pipe_task, pipeline_task, p);
+		schedule_task_config(&p->pipe_task, pipe_desc->priority,
+				     pipe_desc->core);
+	} else {
+		work_init(&p->pipe_work, pipeline_work, p, WORK_HIGH_PRI,
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/audio/pipeline.c]: Right now only focusing on enabling GHW.

File: src/audio/dai.c [hunk 19]
```diff
@@ -170,11 +170,15 @@ static void dai_dma_cb(void *data, uint32_t type, struct dma_sg_elem *next)
 		return;
 	}
 
-	dai_buffer_process(dev);
+	if (!dd->config.timer) {
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/audio/dai.c]: Agreed, but right now focusing on enabling GHW.

File: src/drivers/dw/dma.c (hunk 4 above)
Comment (tlauda) [reply src/drivers/dw/dma.c]: It's the same.

File: src/drivers/dw/dma.c (hunk 7 above)
Comment (tlauda) [reply src/drivers/dw/dma.c]: It's related in the manner, that after removing dw_dma_work function some platforms won't build.

File: src/audio/dai.c [hunk 20]
```diff
@@ -631,18 +647,59 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
 static int dai_copy(struct comp_dev *dev)
 {
 	struct dai_data *dd = comp_get_drvdata(dev);
-	int ret;
+	int ret = 0;
+	int dma_avail_size = 0;
+	int in_size = 0;
+	int out_size = 0;
+	int copy_size = 0;
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/audio/dai.c]: I think according to the rules they should be declared at the top of the function.

File: src/drivers/dw/dma.c (hunk 16 above)
Comment (tlauda) [reply src/drivers/dw/dma.c]: It will pass copied size to the dai.

**Comment by tlauda** [comment]
Review suggestions applied.
I agree, that this makes code very messy and not scalable, but we should postpone any code optimizations after scheduling refactor.

**Comment by mmaka1** [comment]
There are at least three features grouped by the PR:
(1) var size data copying by pipelines,
(2) comp copy operations sequence re-implemented,
(3) timer driver pipeline scheduling

and all are coupled to and enabled by the timer-driven mode.

How about making (1) & (2) common, replacing the old code and make timer-driven vs. dma-int-driven the only difference? That might make the code easier to refactor wrt APIs (as requested) and help to reduce "dead" code (assuming that most of the setups would use either option for the whole topology rather than a mixed mode).

**Comment by lgirdwood** [comment]
@tlauda @mmaka1 we also need to remove work.c and take it's features over into schedule.c. The schedule API needs to be updated to handle synchronous and asynchronous scheduling based on  multiple time sources (e.g. DMA, timers, external IRQs)

**Comment by tlauda** [comment]
@lgirdwood I will see, what I can come up with.

**Comment by lgirdwood** [comment]
@tlauda thanks, I do think will simplify things later on.
